src/request/mod.rs

```diff
@@ -71,73 +69,72 @@ impl<'a, 'b> Debug for Request<'a, 'b> {
-                            SocketAddr::V6(addr6) => format!("{}://[{}]:{}{}", protocol.name(), addr6.ip(), local_addr.port(), path),
-                        }
-                    },
-                    (_, None) => {
-                        return Err("No host specified in request".into())
+    pub fn from_http(req: HttpRequest, local_addr: SocketAddr, protocol: &Protocol)
+                     -> Result<Request, String> {
+        let addr = req.remote_addr().cloned();
+        let (method, uri, version, headers, body) = req.deconstruct();
+        let url = {
+            let path = uri.path();
+            let url_string = match (version, headers.get::<headers::Host>()) {
+                (_, Some(host)) => {
+                    // Attempt to prepend the Host header (mandatory in HTTP/1.1)
+                    if let Some(port) = host.port() {
+                        format!("{}://{}:{}{}", protocol.name(), host.hostname(), port, path)
+                    } else {
+                        format!("{}://{}{}", protocol.name(), host.hostname(), path)
                     }
-                };
-
-                match Url::parse(&url_string) {
-                    Ok(url) => url,
-                    Err(e) => return Err(format!("Couldn't parse requested URL: {}", e))
+                },
+                (v, None) if v < HttpVersion::Http11 => {
+                    // Attempt to use the local address? (host header is not required in HTTP/1.0).
+                    match local_addr {
+                        SocketAddr::V4(addr4) => format!("{}://{}:{}{}", protocol.name(), addr4.ip(), local_addr.port(), path),
+                        SocketAddr::V6(addr6) => format!("{}://[{}]:{}{}", protocol.name(), addr6.ip(), local_addr.port(), path),
+                    }
+                },
+                (_, None) => {
+                    return Err("No host specified in request".into())
                 }
-            },
-            _ => return Err("Unsupported request URI".into())
+            };
+
+            match Url::parse(&url_string) {
+                Ok(url) => url,
+                Err(e) => return Err(format!("Couldn't parse requested URL: {}", e))
+            }
         };
 
         Ok(Request {
             url: url,
             remote_addr: addr,
             local_addr: local_addr,
             headers: headers,
-            body: Body::new(reader),
+            body: Some(body),
             method: method,
             extensions: TypeMap::new(),
             version: version,
             _p: (),
         })
     }
 
+    /// Get the contents of the body as a Vec<u8>
+    ///
+    /// This consumes the body future and turns it into Vec<u8>.  Note this should not be called
+    /// from the main hyper thread, as it will potentially deadlock.
+    pub fn get_body_contents(&mut self) -> Option<Vec<u8>> {
```

Comment:
That sounds great. I don't love the name, but I can't really think of anything better right now. `get_body` is easy to confuse with `.body`, and anything else is as long. Does anyone else have ideas? I'll keep thinking about it.

There is one further issue with this API: it encourages reading an unlimited amount of data over the network, which is basically never what you want to do. The body-parser crate gets around this by having a [default maximum length](https://github.com/iron/body-parser/blob/master/src/lib.rs#L63) and [accepting further configuration](https://github.com/iron/body-parser/blob/master/src/lib.rs#L49) via `extensions`. We could do that here instead, but the API is getting a little complex to go in core.